paasta_tools/kubernetes/bin/paasta_secrets_sync.py

```diff
@@ -215,5 +232,84 @@ def sync_secrets(
     return True
 
 
+def sync_boto_secrets(
+    kube_client: KubeClient,
+    cluster: str,
+    service: str,
+    secret_provider_name: str,
+    vault_cluster_config: Mapping[str, str],
+    soa_dir: str,
+) -> bool:
+    # Update boto key secrets
+    config_loader = PaastaServiceConfigLoader(service=service, soa_dir=soa_dir)
+    for instance_config in config_loader.instance_configs(
+        cluster=cluster, instance_type_class=KubernetesDeploymentConfig
+    ):
+        instance = instance_config.instance
+        boto_keys = instance_config.config_dict.get("boto_keys", [])
+        if not boto_keys:
+            continue
+        boto_keys.sort()
+        secret_data = {}
+        for key in boto_keys:
+            for filetype in ["sh", "yaml", "json", "cfg"]:
+                this_key = key + "." + filetype
+                sanitised_key = this_key.replace(".", "-").replace("_", "--")
+                try:
+                    with open(f"/etc/boto_cfg_private/{this_key}") as f:
+                        secret_data[sanitised_key] = base64.b64encode(
+                            f.read().encode("utf-8")
+                        ).decode("utf-8")
+                except IOError:
+                    log.warning(
+                        f"Boto key {this_key} required for {service} could not be found."
+                    )
+        if not secret_data:
+            continue
+        secret = f"paasta-boto-key-{get_kubernetes_app_name(service, instance)}"
+        hashable_data = "".join([secret_data.get(key, "") for key in secret_data])
+        signature = hashlib.sha1(hashable_data.encode("utf-8")).hexdigest()
+        kubernetes_signature = get_kubernetes_secret_signature(
+            kube_client=kube_client, secret=secret, service=service
+        )
+        if not kubernetes_signature:
+            log.info(f"{secret} for {service} not found, creating")
+            try:
+                create_plaintext_dict_secret(
+                    kube_client=kube_client,
+                    secret_name=secret,
+                    secret_data=secret_data,
+                    service=service,
+                )
+            except ApiException as e:
+                if e.status == 409:
+                    log.warning(f"Secret {secret} for {service} already exists")
+                else:
+                    raise
+            create_kubernetes_secret_signature(
+                kube_client=kube_client,
+                secret=secret,
+                service=service,
+                secret_signature=signature,
+            )
+        elif signature != kubernetes_signature:
+            log.info(f"{secret} for {service} needs updating as signature changed")
```

Comment:
this will cause a service bounce, right? 

do we rotate all keys at once or on a rolling basis? just worried about contant big bounces if this happens all at once :p

Reply:
AFAIK it does not. I tested this by changing the value of a key while I had a shell in the service. Though, I exited the shell right after so there's a possibility it's still doing some delayed bounce.

Reply:
If I did miss this, are you thinking of a particular mechanism that causes a bounce?

Reply:
I *think* that updating the secret content will cause a bounce once setup_kubernetes_job gets to the service in question - but I haven't confirmed that yet

Reply:
https://github.com/Yelp/paasta/blob/c5f92d43a391fc1e12d51a6f3ef6fdbe29f3c30b/paasta_tools/kubernetes_tools.py#L1777 here's a code hint. Maybe we actually will ignore changes if it's just volume based?

Reply:
Generally speaking it would make sense if we just bounced for environment based secrets since I think file based secrets get live updated now in k8s? @Qmando have you thought much about the load when secrets rotate, do they happen all at once, are you satisfied that it's not a huge load on the k8s API? The other area where I'd suggest extra caution for you and reviewers is in the kubernetes_tools code since any case where we generate an invalid k8s object/error/exception could prevent deploying the service(s) and cause a DAR3

Reply:
> def sanitize_for_config_hash
Looks like this is explicitly looking at PAASTA_SECRET_* style secrets, and won't detect changes to the boto secret. But, if `boto_keys` in the config dict changes, it will bounce. This is the behavior we want.